app/src/main/java/com/hedvig/app/feature/marketpicker/MarketPickerActivity.kt

```diff
@@ -0,0 +1,117 @@
+    private val tracker: LanguageSelectionTracker by inject()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        storeBoolean(HAS_SHOWN_MARKET_SELECTION, true)
+
+        var marketAdapter: MarketAdapter? = null
+        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
+
+        marketModel.preselectedMarket.observe(this) { marketString ->
+            marketString?.let {
+                val market = Market.valueOf(marketString)
+                marketAdapter = MarketAdapter(marketModel, market.ordinal)
+                countryList.adapter = marketAdapter
+                countryList.addItemDecoration(
+                    DividerItemDecoration(this, DividerItemDecoration.VERTICAL).apply {
+                        compatDrawable(R.drawable.divider)?.let { setDrawable(it) }
+                    }
+                )
+            }
+
+        }
+        marketModel.loadGeo()
+
+        languageList.adapter = LanguageAdapterNew(languageViewModel, Market.SE)
+        marketModel.selectedMarket.observe(this) { market ->
+            market?.let {
+                languageList.adapter = LanguageAdapterNew(languageViewModel, market)
+            }
+        }
+        languageList.addItemDecoration(
+            DividerItemDecoration(this, DividerItemDecoration.VERTICAL).apply {
+                compatDrawable(R.drawable.divider)?.let { setDrawable(it) }
+            }
+        )
+
+        save.setHapticClickListener {
+            marketAdapter?.let {
+                sharedPreferences.edit()
+                    .putInt(Market.MARKET_SHARED_PREF, marketAdapter!!.getSelectedMarket())
+                    .commit()
+            }
+            val language = languageViewModel.selectedLanguage.value
+            language?.let {
+                setLanguage(language, languageViewModel)
+            }
+            goToMarketingActivity()
+        }
+
+        languageViewModel.selectedLanguage.observe(this) { language ->
+            save.isEnabled = language != null
+        }
+    }
+
+    @SuppressLint("ApplySharedPref") // We want to apply this right away. It's important
+    private fun setLanguage(language: Language, languageViewModel: LanguageViewModel) {
+        PreferenceManager
+            .getDefaultSharedPreferences(this)
+            .edit()
+            .putString(SettingsActivity.SETTING_LANGUAGE, language.toString())
+            .commit()
+
+        language.apply(this)?.let { language ->
+            languageViewModel.updateLanguage(makeLocaleString(language))
+        }
+
+        LocalBroadcastManager
+            .getInstance(this)
+            .sendBroadcast(Intent(LOCALE_BROADCAST))
+    }
+
+    private fun goToMarketingActivity() {
+        startActivity(Intent(this, MarketingActivity::class.java).apply {
+            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
+        })
```

Comment:
Let's introduce a `MarketingActivity.Companion.newInstance(context: Context, withoutHistory: Boolean = false)`-function, where `withoutHistory` would apply the flags listed here. See `NorwegianAuthenticationActivity` for inspiration 🙂